### ISSUE TITLE
[Enhancement] Add the Ability to Use NVIDIA for Docker

### ISSUE THREAD

**Issue by ZaxLofful** [nsarrazin/serge#43]
Allow large language models with graphics card with large RAM.

https://github.com/NVIDIA/nvidia-docker

**Comment by BrunoIsaac27** [comment]
Yes please! the model is very slow on amd epyc cpu..
![image](https://user-images.githubusercontent.com/66482679/228144096-7d684427-ed46-471b-b2aa-51f6c670e2ee.png)

**Comment by fishscene** [comment]
I too would like to see this capability.

**Comment by jerydfans** [comment]
I am willing to test GPU / Nvidia / Docker

**Comment by lucienn3** [comment]
Hey guys,

Please reach out to me if you want to test it.
I have an i9-13900KF, 64 GB DDR5 RAM, and an RTX4090. I would love to help with testing :)

**Comment by gaby** [comment]
This is in the TODO list, we have to split the Dockerfiles first though. Having GPU support will make the image huge. The plan is to publish separate images for GPU/Non-GPU support

**Comment by realies** [comment]
@gaby, any idea when could these be expected? Happy to test too.

**Comment by Jonpro03** [comment]
I got it running. Feel free to borrow: https://gist.github.com/Jonpro03/604430a3e64735a0a9df6b7e385d15be

**Comment by gaby** [comment]
> I got it running. Feel free to borrow: https://gist.github.com/Jonpro03/604430a3e64735a0a9df6b7e385d15be

Ideally during runtime the image should be "-runtime" not "-devel". My plan is to make a compose that use the dockerfile for the final stage

**Comment by gaby** [comment]
> @gaby, any idea when could these be expected? Happy to test too.

This week

**Comment by Jonpro03** [comment]
> > I got it running. Feel free to borrow: https://gist.github.com/Jonpro03/604430a3e64735a0a9df6b7e385d15be
> 
> Ideally during runtime the image should be "-runtime" not "-devel". My plan is to make a compose that use the dockerfile for the final stage

Groovy. Can confirm it works with the runtime image too.

**Comment by gaby** [comment]
@Jonpro03 llama-cpp-python compiles correctly even using the "-runtime" tag?

**Comment by gaby** [comment]
I see that you are adding CMAKE flags, not sure what those do, haha

**Comment by Jonpro03** [comment]
> I see that you are adding CMAKE flags, not sure what those do, haha
Possibly a me problem, since I'm running older CPUs (Xeon E5-2640's), but these flags were necessary to get it to compile with the pip. 
`-DLLAMA_CUBLAS=ON -DLLAMA_NATIVE=ON` should be all that's necessary, but wasn't in my case (I suspect a lama-cpp bug) https://github.com/ggerganov/llama.cpp/issues/1982

**Comment by gaby** [comment]
> > I see that you are adding CMAKE flags, not sure what those do, haha
> > Possibly a me problem, since I'm running older CPUs (Xeon E5-2640's), but these flags were necessary to get it to compile with the pip.
> > `-DLLAMA_CUBLAS=ON -DLLAMA_NATIVE=ON` should be all that's necessary, but wasn't in my case (I suspect a lama-cpp bug) [ggerganov/llama.cpp#1982](https://github.com/ggerganov/llama.cpp/issues/1982)

Thanks for the info!

**Comment by Jonpro03** [comment]
> @Jonpro03 llama-cpp-python compiles correctly even using the "-runtime" tag?

Oops, I spoke too soon. It only works with the devel tag.

**Comment by gaby** [comment]
> > @Jonpro03 llama-cpp-python compiles correctly even using the "-runtime" tag?
> 
> Oops, I spoke too soon. It only works with the devel tag.

Yeah, that's what i figured. The devel tag is needed to compile the llama-cpp-python wheel. After that the runtime tag can be used.

My main has been that the devel tag is like 6.5GB

**Comment by visorcraft** [comment]
We're also willing to test Nvidia support, have 2x RTX 3090 w/nvlink set up on Unraid and eager to Nvidia support working!

**Comment by anethema** [comment]
> > > @Jonpro03 llama-cpp-python compiles correctly even using the "-runtime" tag?
> > 
> > 
> > Oops, I spoke too soon. It only works with the devel tag.
> 
> Yeah, that's what i figured. The devel tag is needed to compile the llama-cpp-python wheel. After that the runtime tag can be used.
> 
> My main problem has been that the devel tag is like 6.5GB

Any progress? I would love to be able to run a local LLM on unraid with GPU acceleration

**Comment by gaby** [comment]
> > > > @Jonpro03 llama-cpp-python compiles correctly even using the "-runtime" tag?
> > > 
> > > 
> > > Oops, I spoke too soon. It only works with the devel tag.
> > 
> > 
> > Yeah, that's what i figured. The devel tag is needed to compile the llama-cpp-python wheel. After that the runtime tag can be used.
> > My main problem has been that the devel tag is like 6.5GB
> 
> Any progress? I would love to be able to run a local LLM on unraid with GPU acceleration

I'm on vacation, will get this done in like 2 weeks. :-)

**Comment by creed2415** [comment]
![企业微信截图_16990015161142](https://github.com/serge-chat/serge/assets/147784107/c5e19839-929f-416c-bd82-0ccb0bf5c5a1)

I  just found no use of gpu when I started a chat ( 30 GPU Layers) ,Should I install CUDA inside Docker?

Below are my proposed modifications:
step1: set the LLAMA_CUBLAS=1 environment variable
```
# Set ENV
ENV NODE_ENV='production'
ENV TZ=Etc/UTC
ENV LLAMA_CUBLAS=1
WORKDIR /usr/src/app
```
step2: replace the installation method for "llama-cpp-python" in the "/scripts/deploy.sh" file

```
# Install python bindings
UNAME_M=$(dpkg --print-architecture) CMAKE_ARGS="-DLLAMA_CUBLAS=on" pip install llama-cpp-python || {
	echo 'Failed to install llama-cpp-python'
	exit 1
}
```
step3:install  nvidia-container-toolkit

**Comment by SurvivaLlama** [comment]
@creed2415 Did you get GPU to work?

**Comment by syonfox** [comment]
Cool stuff does anyone have numbers on the infrance speed on gpu vs cpu using this

**Comment by creed2415** [comment]
> @creed2415 Did you get GPU to work?

yes! After I installed CUDA inside Docker,it worked with  gpu.
![企业微信截图_16998546303524](https://github.com/serge-chat/serge/assets/147784107/a97f55a1-61c8-4c22-8552-d592950630e6)

**Comment by anethema** [comment]
@creed2415 did you do this in unraid by chance? Not sure how to actually make those edits to the package since it comes from the unraid community applications. 

@gaby any tips on how to make that happen? Or is that something you still planned on ?

**Comment by gaby** [comment]
@anethema I have it ready, but I need to merge and test #866 first :-)

**Comment by anethema** [comment]
> @anethema I have it ready, but I need to merge and test #866 first :-)

Awesome! Exciting to see.

**Comment by creed2415** [comment]
@gaby If possible, I would like to submit my solution (linux for nvidia) to the repository.

**Comment by gaby** [comment]
@creed2415 Serge has to work for both gpu/non-gpu, so images have to be separate, etc. So, it's a bit more complicated than just adding nvidia-toolkit. If you got a solution that works for everyone, sure.

**Comment by creed2415** [comment]
@gaby well,here are the adjustments I made. just take a look and let me know if they are useful.
https://github.com/creed2415/serge/tree/gpu-support

**Comment by anethema** [comment]
> @mkeshav This week :-)

Any luck with this?

**Comment by gaby** [comment]
> > @mkeshav This week :-)
> 
> Any luck with this?

Still in progress

**Comment by iamzoltan** [comment]
Can I help?

**Comment by TheQuickestFox** [comment]
@gaby looks like #944 passed checks, can I get excited? thanks so much for the amazing work on the rest of this by the way - by far the most user friendly AI interface for local LLMs on Unraid.

**Comment by gaby** [comment]
@TheQuickestFox Not yet, i'm helping the `llama-cpp-python` folks tests their new wheels to make this more straight forward.

**Comment by JuniperChris929** [comment]
any ETA on this? and also on the latest IPv6 fixes? The last release is from Feb :(

**Comment by gaby** [comment]
@JuniperChris929 Within the next 2 weeks. The main library used by Serge now has built-in support for CUDA Python wheels which will streamline our approach to support multiple platforms.

Even though last release was in February, there's over 100 commits between then and now.

**Comment by CodeMazeSolver** [comment]
Are there any updates regarding Serge using GPU acceleration?

**Comment by VendorAttestation** [comment]
@gaby is this gunna happen?

**Comment by Imperial929** [comment]
Any news here? We keep hearing over and over again "within a week or two" but the last release is from February and neither CUDA SUpport is easily working nor support for IPv6-only :(

**Comment by warlordattack** [comment]
also interested for nvidia gpu, having RTX 3070 in my proxmox/mint/docker VM 
:)

**Comment by SurvivaLlama** [comment]
at this point just move on and use ollama to run your llms.

**Comment by gaby** [comment]
Yeah, I suggest using https://github.com/open-webui/open-webui
